File: java/arcade/DifferentSquares.java
public final class DifferentSquares {

	int differentSquares(int[][] matrix) {
	    int h = matrix.length, w = matrix[0].length;
	    ArrayList<String> sums = new ArrayList();
	    String s = "";
	    for(int i = 1; i < h; i++) {
	        for(int j = 1; j < w; j++, s = "") {
	            s += ""+matrix[i][j]+matrix[i][j - 1]+
	                    matrix[i - 1][j - 1]+matrix[i - 1][j];
	            if(!sums.contains(s)) { sums.add(s); }
	        }
	    }
	    return sums.size();
	}
}

File: ruby/arcade/differentSquares.rb
def differentSquares(matrix)
    str, sums = "", []
    (1...matrix.size).each do |i|
        (1...matrix[0].size).each do |j|
            str += matrix[i][j].to_s+matrix[i][j - 1].to_s+
	            matrix[i - 1][j - 1].to_s+matrix[i - 1][j].to_s
            sums.push str unless sums.include? str
            str = ""
        end
    end
    return sums.size
end